### PR TITLE
GCP: populate encryption values into tf and machines

### PR DESCRIPTION
**Pull request by michaelgugino**
This commit ensures encryption values are properly
populated into terraform and machines/machinesets.

### CODE REVIEW

File: pkg/asset/machines/gcp/machines.go [hunk 1]
```diff
@@ -77,6 +77,22 @@ func provider(clusterID string, platform *gcp.Platform, mpool *gcp.MachinePool,
 		return nil, err
 	}
 
+	var encryptionKey *gcpprovider.GCPEncryptionKeyReference
+
+	if mpool.OSDisk.EncryptionKey != nil {
```
Comment (michaelgugino) [inline pkg/asset/machines/gcp/machines.go:82]: Is there a reason that OSDisk.EncryptionKey can't just be an imported type from the GCP provider?  That should allow us to a do a straight assignment here.

File: pkg/tfvars/gcp/gcp.go [hunk 2]
```diff
@@ -2,12 +2,17 @@ package gcp
...
 import (
 	"encoding/json"
+	"fmt"
 
 	gcpprovider "github.com/openshift/cluster-api-provider-gcp/pkg/apis/gcpprovider/v1beta1"
 
 	"github.com/openshift/installer/pkg/types"
 )
 
+const (
```
Comment (michaelgugino) [inline pkg/tfvars/gcp/gcp.go:12]: Maybe this belongs in GCP provider in a util package?

File: pkg/tfvars/gcp/gcp.go [hunk 3]
```diff
@@ -78,5 +84,17 @@ func TFVars(sources TFVarsSources) ([]byte, error) {
...
 	}
 
+	if masterConfig.Disks[0].EncryptionKey != nil {
+		cfg.VolumeKMSKeyLink = generateDiskEncryptionKeyLink(masterConfig.Disks[0].EncryptionKey, masterConfig.ProjectID)
+	}
+
 	return json.MarshalIndent(cfg, "", "  ")
 }
+
+func generateDiskEncryptionKeyLink(keyRef *gcpprovider.GCPEncryptionKeyReference, projectID string) string {
```
Comment (michaelgugino) [inline pkg/tfvars/gcp/gcp.go:94]: Maybe this belongs in GCP provider in a util package?

File: pkg/asset/machines/gcp/machines.go (hunk 1 above)
Comment (patrickdillon) [reply pkg/asset/machines/gcp/machines.go:82]: Yeah. `OSDisk.EncryptionKey` is defined in the types package, which represents the installer API, so we try to keep it free of external dependencies considering other projects import the types package.

File: pkg/asset/machines/gcp/machines.go [hunk 4]
```diff
@@ -77,6 +77,22 @@ func provider(clusterID string, platform *gcp.Platform, mpool *gcp.MachinePool,
...
+				Name:      mpool.OSDisk.EncryptionKey.KMSKey.Name,
+				KeyRing:   mpool.OSDisk.EncryptionKey.KMSKey.KeyRing,
+				ProjectID: mpool.OSDisk.EncryptionKey.KMSKey.ProjectID,
+				Location:  mpool.OSDisk.EncryptionKey.KMSKey.Location,
+			},
+			KMSKeyServiceAccount: mpool.OSDisk.EncryptionKey.KMSKeyServiceAccount,
+		}
+	} else {
+		encryptionKey = nil
+	}
```
Comment (staebler) [inline pkg/asset/machines/gcp/machines.go:94]: The else block is unnecessary.

File: pkg/asset/machines/gcp/machines.go [hunk 5]
```diff
@@ -77,6 +77,22 @@ func provider(clusterID string, platform *gcp.Platform, mpool *gcp.MachinePool,
...
+			KMSKey: &gcpprovider.GCPKMSKeyReference{
+				Name:      mpool.OSDisk.EncryptionKey.KMSKey.Name,
+				KeyRing:   mpool.OSDisk.EncryptionKey.KMSKey.KeyRing,
+				ProjectID: mpool.OSDisk.EncryptionKey.KMSKey.ProjectID,
+				Location:  mpool.OSDisk.EncryptionKey.KMSKey.Location,
+			},
+			KMSKeyServiceAccount: mpool.OSDisk.EncryptionKey.KMSKeyServiceAccount,
+		}
+	} else {
+		encryptionKey = nil
```
Comment (patrickdillon) [inline pkg/asset/machines/gcp/machines.go:93]: nit: not sure this is necessary (should already be nil). if you want to leave it in, moving it out of `else` and before `if` would be more concise.

File: data/data/gcp/bootstrap/variables.tf [hunk 6]
```diff
@@ -59,6 +59,12 @@ variable "root_volume_type" {
   description = "The volume type for the bootstrap node's root volume."
 }
 
+variable "root_volume_kms_key_link" {
+  type       = string
+  description = "The GCP self link of KMS key to encrypt the volume"
```
Comment (patrickdillon) [inline data/data/gcp/bootstrap/variables.tf:64]: nit: missing period.

File: data/data/gcp/master/variables.tf [hunk 7]
```diff
@@ -44,6 +44,12 @@ variable "root_volume_type" {
   description = "The type of volume for the root block device."
 }
 
+variable "root_volume_kms_key_link" {
+  type       = string
+  description = "The GCP self link of KMS key to encrypt the volume"
```
Comment (patrickdillon) [inline data/data/gcp/master/variables.tf:49]: nit: missing period

File: data/data/gcp/variables-gcp.tf [hunk 8]
```diff
@@ -115,3 +115,8 @@ variable "gcp_image_licenses" {
   default = []
 }
 
+variable "gcp_root_volume_kms_key_link" {
+  type       = string
+  description = "The GCP self link of KMS key to encrypt the volume"
```
Comment (patrickdillon) [inline data/data/gcp/variables-gcp.tf:120]: nit: missing period

File: data/data/gcp/main.tf [hunk 9]
```diff
@@ -47,8 +47,9 @@ module "master" {
   subnet         = module.network.master_subnet
   zones          = distinct(var.gcp_master_availability_zones)
 
-  root_volume_size = var.gcp_master_root_volume_size
-  root_volume_type = var.gcp_master_root_volume_type
+  root_volume_size         = var.gcp_master_root_volume_size
+  root_volume_type         = var.gcp_master_root_volume_type
+  root_volume_kms_key_link = var.gcp_root_volume_kms_key_link
```
Comment (staebler) [inline data/data/gcp/main.tf:52]: Need to add this to the "bootstrap" module, too.

https://github.com/openshift/installer/pull/4397/files#diff-c9542c57acc94c4d250fd8d806ac9cf57c9ccd52c3c8e282a34527198949ff57R35

**Review by staebler — COMMENTED** [review]
Looks good

File: pkg/tfvars/gcp/gcp.go (hunk 3 above)
Comment (patrickdillon) [reply pkg/tfvars/gcp/gcp.go:94]: To echo our conversation on slack, I would handle this in terraform:

1. create a type for the key similar to Auth in TFVars and pass the values in that way
2. conditionally create the self link variable (or null value therein) with logic similar to BYO networking: https://github.com/openshift/installer/blob/master/data/data/gcp/network/common.tf

I'm also fine with this approach and think it's OK to have this logic live here.

**Review by patrickdillon — COMMENTED** [review]
Left a few comments, but this looks really good to me.

File: data/data/gcp/master/variables.tf [hunk 10]
```diff
@@ -44,6 +44,12 @@ variable "root_volume_type" {
...
+  default = null
```
Comment (patrickdillon) [inline data/data/gcp/master/variables.tf:50]: column alignment is off and causing `tf-fmt` to fail

File: data/data/gcp/variables-gcp.tf [hunk 11]
```diff
@@ -115,3 +115,8 @@ variable "gcp_image_licenses" {
...
+  default = null
```
Comment (patrickdillon) [inline data/data/gcp/variables-gcp.tf:121]: column alignment is off and causing `tf-fmt` to fail

File: data/data/gcp/bootstrap/variables.tf [hunk 12]
```diff
@@ -59,6 +59,12 @@ variable "root_volume_type" {
...
+  default = null
```
Comment (patrickdillon) [inline data/data/gcp/bootstrap/variables.tf:65]: column alignment is off and causing `tf-fmt` to fail

File: data/data/gcp/variables-gcp.tf (hunk 11 above)
Comment (patrickdillon) [reply data/data/gcp/variables-gcp.tf:121]: > column alignment is off and causing `tf-fmt` to fail

weird, tf-fmt seems to be suggesting different alignment for this file than the other two:
https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/4397/pull-ci-openshift-installer-master-tf-fmt/1329497726143959040/build-log.txt

**Comment by patrickdillon** [comment]
I would pull the missing dependencies into this PR in two separate commit: one commit for your changes and a separate one for the vendored files.

File: data/data/gcp/main.tf (hunk 9 above)
Comment (JoelSpeed) [reply data/data/gcp/main.tf:52]: Is that not done as the change in `data/data/gcp/bootstrap/main.tf`, or does that need to go into `data/data/bootstrap` somewhere?

File: pkg/asset/machines/gcp/machines.go (hunk 5 above)
Comment (JoelSpeed) [reply pkg/asset/machines/gcp/machines.go:93]: This shouldn't be needed as `var encryptionKey *gcpprovider.GCPEncryptionKeyReference` above the if makes it a nil anyway

File: data/data/gcp/main.tf (hunk 9 above)
Comment (patrickdillon) [reply data/data/gcp/main.tf:52]: In `/data/data/gcp/main.tf` as linked above. It is currently only passing the variable from main to master, but needs to do the same for bootstrap.

**Comment by JoelSpeed** [comment]
@patrickdillon I've fixed all of the comments that were left (I believe) on this branch https://github.com/openshift/installer/compare/master...JoelSpeed:gcp-encrypt

I've also added a commit for the missing dependencies as requested.

I didn't do anything for this one (https://github.com/openshift/installer/pull/4397/files#r527187850) as it seemed there wasn't a strong preference to change it.

Mike is out today and has suggested he will be out all week, I wondered if you could give my branch a check over and, if that's looking good, I could squash my fixups into his commit, open a new PR and we can merge that one while he is out, WDYT? Trying to give this PR as much time to soak in QE as possible

**Comment by patrickdillon** [comment]
> @patrickdillon I've fixed all of the comments that were left (I believe) on this branch [master...JoelSpeed:gcp-encrypt](https://github.com/openshift/installer/compare/master...JoelSpeed:gcp-encrypt)
> 
> I've also added a commit for the missing dependencies as requested.
> 
> I didn't do anything for this one (https://github.com/openshift/installer/pull/4397/files#r527187850) as it seemed there wasn't a strong preference to change it.
> 
> Mike is out today and has suggested he will be out all week, I wondered if you could give my branch a check over and, if that's looking good, I could squash my fixups into his commit, open a new PR and we can merge that one while he is out, WDYT? Trying to give this PR as much time to soak in QE as possible

The branch in general looks good. I don't have approval rights at this level, so I can't push this through. @staebler is out until Monday. I would recommend just opening a PR and getting everything to pass so it could be ready to go on Monday. 

If you think it's really important to get it in ASAP we can try to coordinate an approve with @sdodson 's help.

**Comment by staebler** [comment]
@michaelgugino The dependency mess has finally merged. This PR needs another rebase.

**Review by staebler — COMMENTED** [review]
/lgtm
/approve
/label tide/merge-method-squash

**Comment by staebler** [comment]
The GCP install succeeded in the latest e2e-gcp test.

**Comment by staebler** [comment]
All the changes in this PR are in GCP-specific files.

/override ci/prow/e2e-aws

**Pull request merged**